File: 285-inorder-successor-in-bst/285-inorder-successor-in-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderSuccessor(self, root: TreeNode, p: TreeNode) -> Optional[TreeNode]:
        prev = None
        stack = deque()
        while True:
            while root:
                stack.append(root)
                root = root.left
            if not stack:
                return None
            root = stack.pop()
            if p == prev:
                return root
            prev = root
            root = root.right

File: 285-inorder-successor-in-bst/285-inorder-successor-in-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        TreeNode prev = null;
        Deque<TreeNode> stack = new ArrayDeque<>();
        while(true){
            while(root != null) {
                stack.push(root);
                root = root.left;
            }
            if(stack.isEmpty()) return null;
            root = stack.pop();
            if(p.equals(prev)) return root;
            prev = root;
            root = root.right;
        }        
    }
}